src/librustc_mir/build/matches/test.rs

```diff
@@ -258,69 +256,77 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
-                            .and_then(|t| t.ty.builtin_index())
-                            .map_or(false, |t| t == self.hir.tcx().types.u8);
-                        if is_array_ptr {
-                            self.hir
-                                .tcx()
-                                .interpret_interner
-                                .borrow()
-                                .get_alloc(p.alloc_id)
-                                .map(|alloc| &alloc.bytes[..])
-                        } else {
-                            None
-                        }
-                    },
-                    _ => None,
-                };
-                // If we're using b"..." as a pattern, we need to insert an
-                // unsizing coercion, as the byte string has the type &[u8; N].
-                let expect = if let Some(bytes) = bytes {
-                    let tcx = self.hir.tcx();
-
-                    // Unsize the place to &[u8], too, if necessary.
-                    if let ty::TyRef(region, mt) = ty.sty {
-                        if let ty::TyArray(_, _) = mt.ty.sty {
-                            ty = tcx.mk_imm_ref(region, tcx.mk_slice(tcx.types.u8));
-                            let val_slice = self.temp(ty, test.span);
-                            self.cfg.push_assign(block, source_info, &val_slice,
-                                                 Rvalue::Cast(CastKind::Unsize, val, ty));
-                            val = Operand::Move(val_slice);
-                        }
-                    }
-
-                    assert!(ty.is_slice());
-
-                    let array_ty = tcx.mk_array(tcx.types.u8, bytes.len() as u64);
-                    let array_ref = tcx.mk_imm_ref(tcx.types.re_static, array_ty);
-                    let array = self.literal_operand(test.span, array_ref, Literal::Value {
-                        value
-                    });
-
-                    let slice = self.temp(ty, test.span);
-                    self.cfg.push_assign(block, source_info, &slice,
-                                         Rvalue::Cast(CastKind::Unsize, array, ty));
-                    Operand::Move(slice)
-                } else {
-                    self.literal_operand(test.span, ty, Literal::Value {
-                        value
-                    })
-                };
-
-                // Use PartialEq::eq for references, instead of BinOp::Eq.
+                let mut expect = self.literal_operand(test.span, ty, Literal::Value {
+                    value
+                });
+                // Use PartialEq::eq instead of BinOp::Eq
+                // (the binop can only handle primitives)
                 let fail = self.cfg.start_new_block();
-                let str_or_bytestr = ty
-                    .builtin_deref(true, ty::NoPreference)
-                    .map(|tam| tam.ty);
-                if let Some(ty) = str_or_bytestr {
+                if !ty.is_scalar() {
```

Comment:
@eddyb is this what you had in mind with not treating references special? Any non-scalar will be copied into a local (so we can take a reference to it, since `Eq::eq` takes references) and then compared with its `Eq` impl.